### PR TITLE
Enabled multiple GMT/GMX files for GSEA

### PR DESCRIPTION
**Pull request by WackerO**
Solves #102, the additional files generate subtabs as in this example report
[GSE50790_multigmt2.html.zip](https://github.com/nf-core/differentialabundance/files/11091139/GSE50790_multigmt2.html.zip)

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/differentialabundance/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/differentialabundance _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by pinin4fjords** [comment]
I was waiting on https://github.com/nf-core/modules/pull/3144 to suggest improvements- will PR your feature branch later :-)

File: nextflow_schema.json [hunk 1]
```diff
@@ -778,7 +778,7 @@
                 "gsea_gene_sets": {
                     "type": "string",
                     "default": "None",
-                    "description": "Gene sets in GMT or GMX-format",
+                    "description": "Gene sets in GMT or GMX-format (multiple input files are possible)",
```
Comment (pinin4fjords) [inline nextflow_schema.json:781]:
```suggestion
                    "description": "Gene sets in GMT or GMX-format (multiple comma-separated input files are possible)",
```

File: workflows/differentialabundance.nf [hunk 2]
```diff
@@ -38,11 +38,12 @@ if (params.study_type == 'affy_array'){
...
 if (params.gsea_run) { 
     if (params.gsea_gene_sets){
-        gene_sets_file = file(params.gsea_gene_sets, checkIfExists: true)
+        gene_sets_files = params.gsea_gene_sets.split(" ")
+        gene_sets_files = gene_sets_files.collect { file(it, checkIfExists: true) }
     } else {
         error("GSEA activated but gene set file not specified!")
     }
-} else { gene_sets_file = [] } 
+} else { gene_sets_files = [] }
```
Comment (pinin4fjords) [inline workflows/differentialabundance.nf:46]:
```suggestion
} 
```
This variable isn't even used if the gene sets functionality is deactivated, so we can probably scrap it

File: workflows/differentialabundance.nf [hunk 3]
```diff
@@ -38,11 +38,12 @@ if (params.study_type == 'affy_array'){
 if (params.control_features) { ch_control_features = file(params.control_features, checkIfExists: true) } else { ch_control_features = [[],[]] } 
 if (params.gsea_run) { 
     if (params.gsea_gene_sets){
-        gene_sets_file = file(params.gsea_gene_sets, checkIfExists: true)
+        gene_sets_files = params.gsea_gene_sets.split(" ")
+        gene_sets_files = gene_sets_files.collect { file(it, checkIfExists: true) }
```
Comment (pinin4fjords) [inline workflows/differentialabundance.nf:42]:
```suggestion
        gene_sets_files = params.gsea_gene_sets.split(",")
        ch_gene_sets = Channel.of(gene_sets_files).map { file(it, checkIfExists: true) }
```
Let's just shortcut to the channel

File: workflows/differentialabundance.nf [hunk 4]
```diff
@@ -313,15 +312,25 @@ workflow DIFFERENTIALABUNDANCE {
             VALIDATOR.out.feature_meta.map{ it[1] },
             [params.features_id_col, params.features_name_col]    
         )
-    
+
+        // Make channel from current file and get it's name for correct saving of output
+
+        ch_gene_sets = Channel.fromList(gene_sets_files)
+
```
Comment (pinin4fjords) [inline workflows/differentialabundance.nf:319]:
```suggestion
```
Hopefully won't need this with the above

File: conf/modules.config [hunk 5]
```diff
@@ -190,15 +190,15 @@ process {
     }
 
     withName: GSEA_GSEA {
-        ext.prefix = { "${meta.id}." }
+        ext.prefix = { "${meta.id}.${meta.gmt_name}." }
         publishDir = [
             [
-                path: { "${params.outdir}/tables/gsea/${meta.id}" },
+                path: { "${params.outdir}/tables/gsea/${meta.id}/${meta.gmt_name}" },
```
Comment (pinin4fjords) [inline conf/modules.config:196]:
```suggestion
                path: { "${params.outdir}/tables/gsea/${meta.id}/${gene_sets.baseName}" },
```
I don't think we actually need to go through the meta for this

File: conf/modules.config [hunk 6]
```diff
@@ -190,15 +190,15 @@ process {
...
                 mode: params.publish_dir_mode,
                 pattern: '*gsea_report_for_*.tsv'
             ],
             [
-                path: { "${params.outdir}/plots/gsea/${meta.id}" },
+                path: { "${params.outdir}/plots/gsea/${meta.id}/${meta.gmt_name}" },
```
Comment (pinin4fjords) [inline conf/modules.config:201]:
```suggestion
                path: { "${params.outdir}/plots/gsea/${meta.id}/${gene_sets.baseName}" },
```

File: workflows/differentialabundance.nf [hunk 7]
```diff
@@ -313,15 +312,25 @@ workflow DIFFERENTIALABUNDANCE {
...
         ch_gsea_inputs = CUSTOM_TABULARTOGSEAGCT.out.gct
             .map{ it.tail() }
             .combine(CUSTOM_TABULARTOGSEACLS.out.cls)
             .map{ tuple(it[1], it[0], it[2]) }
             .combine(ch_gene_sets)
+            .map{ tuple( ["gmt_name": it[3].getName().split("\\.")[0..-2].join("."),
+                            "reference": it[0].reference,
+                            "target": it[0].target,
+                            "id": it[0].id],
+                        it[1], it[2], it[3]) }
```
Comment (pinin4fjords) [inline workflows/differentialabundance.nf:333]:
```suggestion
```
Apologies, but I think we can lose this entirely, it's only used for the modules.config, which I _think_ can reference its gene_sets input directly

File: assets/differentialabundance_report.Rmd [hunk 8]
```diff
@@ -815,17 +815,23 @@ if (any(unlist(params[paste0(possible_gene_set_methods, '_run')]))){
...
-
-      for (i in 1:nrow(contrasts)){
-        cat("\n#### ", contrast_descriptions[i], "\n")
+      for (gmt_file in strsplit(params$gsea_gene_sets, " ")[[1]]) {
+        gmt_name <- strsplit(basename(gmt_file), '.', fixed=T)[[1]]
+        gmt_name <- paste0(head(gmt_name, -1), collapse=".")
         
-        target_gsea_results <- read_metadata(target_gsea_tables[i])[,c(-2,-3)]
-        print( htmltools::tagList(datatable(target_gsea_results, caption = paste0("\nTarget (", contrasts$target[i], ")\n"), rownames = FALSE) ))
+        cat("\n#### ", gmt_name ," {.tabset}\n")
```
Comment (pinin4fjords) [inline assets/differentialabundance_report.Rmd:822]: I'm not 100% sure about the hierarchical tabsets, I find the display a bit confusing, but let's go with it for now.

File: assets/differentialabundance_report.Rmd [hunk 9]
```diff
@@ -815,17 +815,23 @@ if (any(unlist(params[paste0(possible_gene_set_methods, '_run')]))){
     if (unlist(params[paste0(gene_set_method, '_run')])){
       cat("\n### ", toupper(gene_set_method) ," {.tabset}\n")
       
-      reference_gsea_tables <- paste0(contrasts$id, '.gsea_report_for_', contrasts$reference, '.tsv')
-      target_gsea_tables <- paste0(contrasts$id, '.gsea_report_for_', contrasts$target, '.tsv')
-
-      for (i in 1:nrow(contrasts)){
-        cat("\n#### ", contrast_descriptions[i], "\n")
+      for (gmt_file in strsplit(params$gsea_gene_sets, " ")[[1]]) {
```
Comment (pinin4fjords) [inline assets/differentialabundance_report.Rmd:818]:
```suggestion
      for (gmt_file in simpleSplit(params$gsea_gene_sets)) {
```
(just a shortcut function that's available to this file via shinyngs)

File: assets/differentialabundance_report.Rmd [hunk 10]
```diff
@@ -815,17 +815,23 @@ if (any(unlist(params[paste0(possible_gene_set_methods, '_run')]))){
...
+        gmt_name <- strsplit(basename(gmt_file), '.', fixed=T)[[1]]
+        gmt_name <- paste0(head(gmt_name, -1), collapse=".")
```
Comment (pinin4fjords) [inline assets/differentialabundance_report.Rmd:820]:
```suggestion
        gmt_name <- basename(gmt_file)
```

Sorry, I'm not sure what you were trying to do there, but it seems like a plain `basename` does the same thing

**Review by pinin4fjords — COMMENTED** [review]
So I changed my mind- let's just go with your delimited string for now- but comma-separated will be easier to work with. 

I think we can simplify things a bit though - hence the suggestions.

File: assets/differentialabundance_report.Rmd (hunk 8 above)
Comment (WackerO) [reply assets/differentialabundance_report.Rmd:822]: I know what you mean but dunno what would be a better alternative

File: assets/differentialabundance_report.Rmd (hunk 10 above)
Comment (WackerO) [reply assets/differentialabundance_report.Rmd:820]: Hmm, that was to remove the file extension. I suppose we can leave it in though

**Comment by pinin4fjords** [comment]
Also don't forget to update the CHANGELOG (could you also add yourself as reviewer on the last one, I forgot!)

File: assets/differentialabundance_report.Rmd (hunk 10 above)
Comment (pinin4fjords) [reply assets/differentialabundance_report.Rmd]:
```
> basename(tools::file_path_sans_ext('s3://fee/fi/fo.txt.fi'))
[1] "fo.txt"
```
?

**Review by pinin4fjords — APPROVED** [review]
LGTM- thanks for the work!

**Pull request merged**